### PR TITLE
Implement mixed index aggregations

### LINKED ISSUE #3202: Add strategy that will optimize numeric aggregations
The PR #2200 add a strategy that asks the mixed index to count the number of matching elements. This optimization can be improved and add aggregation functions on numeric fields (`min()`, `max()`, `mean()` and `sum()`).

The strategy can optimize queries that look like:
 - `g.V().has(...).values('num_field').min()`
 - `g.V().has(...).values('num_field').max()`
 - `g.V().has(...).values('num_field').avg()`
 - `g.V().has(...).values('num_field').sum()`

The storage backend is then not touched and the performance should be highly increased.

### PR DESCRIPTION
**Pull request by To-om**
Fixes #3202

-----

Thank you for contributing to JanusGraph!

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there an issue associated with this PR? Is it referenced in the commit message?
- [x] Does your PR body contain #xyz where xyz is the issue number you are trying to resolve?
- [x] Has your PR been rebased against the latest commit within the target branch (typically `master`)?
- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [x] Have you written and/or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](https://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE.txt file, including the main LICENSE.txt file in the root of this repository?
- [ ] If applicable, have you updated the NOTICE.txt file, including the main NOTICE.txt file found in the root of this repository?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### CODE REVIEW

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/strategy/JanusGraphMixedIndexAggStrategy.java
Comment (porunov) [inline janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/strategy/JanusGraphMixedIndexAggStrategy.java:41]: `count` > `aggregation`

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/strategy/JanusGraphMixedIndexAggStrategy.java
Comment (porunov) [inline janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/strategy/JanusGraphMixedIndexAggStrategy.java:116]: `count` > `aggregation`

File: janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/rest/RestElasticSearchClient.java
Comment (porunov) [inline janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/rest/RestElasticSearchClient.java:430]: I think we should also add documentation about aggregation loose of precision for big numbers when using ElasticSearch. As noted here, for any number more then `2^53` the result is approximate: https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations.html#limits-for-long-values

I think users should now about this limitation and if they want precise values they need to disable this optimization in their queries. I.e.:
```
g.traversal().withoutStrategies(JanusGraphMixedIndexAggStrategy.class).has("someProperty", 123).sum()
```

**Review by porunov — COMMENTED** [review]
Thank you for this feature contribution @To-om !
I glanced over the PR very quickly and found some Javadoc are using `count query` where the logic is changed for `aggregation query`. I also think we need to add doc explaining a precision problem when using ElasticSearch for aggregation.
I didn't check `JanusGraphMixedIndexAggStep` and `JanusGraphMixedIndexAggStrategy` too deep yet, but hopefully @li-boxuan could take a look at these two classes because we was adding the count optimization strategy previously.
I also didn't check tests and Solr with Lucene implementations yet but will take a look at them soon.
I left small nitpicks below and will continue the review soon

File: janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/rest/RestElasticSearchClient.java
Comment (porunov) [reply janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/rest/RestElasticSearchClient.java:430]: I'm also thinking that maybe it would make sense to actually keep Count aggregation strategy separate from other aggregations optimization strategy. If so, users might want to opt-in using and accept aggregation limits it or not using it and to not have any limits.

Comment (li-boxuan) [reply janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/rest/RestElasticSearchClient.java:430]: > I'm also thinking that maybe it would make sense to actually keep Count aggregation strategy separate from other aggregations optimization strategy

That makes sense because even ES count (without aggregation) itself is an approximate...

File: janusgraph-lucene/src/main/java/org/janusgraph/diskstorage/lucene/LuceneIndex.java
Comment (li-boxuan) [inline janusgraph-lucene/src/main/java/org/janusgraph/diskstorage/lucene/LuceneIndex.java:954]: Use `UnsupportedOperationException`

**Review by li-boxuan — COMMENTED** [review]
Looks great! We also need to:
1) Add a changelog explaining that `JanusGraphMixedIndexCountStrategy` has been renamed `JanusGraphMixedIndexAggStrategy` since it's a breaking change.
2) (Optional) Make this strategy configurable as suggested [here](https://github.com/JanusGraph/janusgraph/pull/3203#discussion_r979960051). We can add it afterward if you don't have time.

File: janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/rest/RestElasticSearchClient.java
Comment (porunov) [reply janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/rest/RestElasticSearchClient.java:430]: > That makes sense because even ES count (without aggregation) itself is an approximate...

I think you are referring to the old implementation (probably in 0.2 or 0.3 version of JanusGraph) where we were using [total hits](https://www.elastic.co/guide/en/elasticsearch/reference/7.17/search-your-data.html#track-total-hits).

Right now for ElasticSearch implementation we are using [Count API](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-count.html) which suppose to return accurate value unless I'm missing something. At least I didn't find information that it's not accurate.

So, my point is that users most likely want to have `count` optimization in their queries all the times because it's accurate but aggregation optimization isn't accurate when the resulting value is more then 2^53. So, some users may want to exclude aggregation optimization in some cases but have count optimization enabled in their queries.

That's why I suggested to have both optimizations for `count` and for `aggregation`.

Another reason is that:
`g.V().has("foo", "bar").limit(1000).count()` - can use optimization with a specific logic like `Math.min(limit, countResultFromIndex)` but I can't say the same for other aggregation queries.

Moreover, later (not necessary in this PR) I think we will need to add a configuration parameter which excludes this aggregation optimization by default - otherwise it may be a breaking change for some users due to approximate results in some cases.

Comment (li-boxuan) [reply janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/rest/RestElasticSearchClient.java:430]: > Right now for ElasticSearch implementation we are using [Count API](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-count.html) which suppose to return accurate value unless I'm missing something. At least I didn't find information that it's not accurate.

You are right! Never mind, I misremembered it :)

Comment (To-om) [reply janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/rest/RestElasticSearchClient.java:430]: ElasticSearch count API is accurate but the Lucene one is not. It only counts the 1000 first elements ([TotalHits](https://lucene.apache.org/core/8_11_2/core/org/apache/lucene/search/TotalHits.html)). It was not the case in Lucene 7.7.2 (JanusGraph 5.x).

I agree with you, accurate optimizations should not be mixed with approximate one. I can split it.

File: janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/rest/RestElasticSearchClient.java
Comment (To-om) [reply janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/rest/RestElasticSearchClient.java:466]: are you sure ?

Comment (li-boxuan) [reply janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/rest/RestElasticSearchClient.java:466]: I mean, you commented out line 454 already, then why not just remove it?

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/step/JanusGraphMixedIndexAggStep.java
Comment (li-boxuan) [inline janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/step/JanusGraphMixedIndexAggStep.java:86]: Maybe I am missing something, but is this branch covered by any test case? I didn't find any but not sure why codecov didn't complain.

**Review by li-boxuan — COMMENTED** [review]
Thanks for the refactoring! Could you please also add a changelog/user doc to briefly introduce `JanusGraphMixedIndexAggStrategy` so that people know how to disable it when they need accurate counts?

File: janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphIndexTest.java
Comment (li-boxuan) [inline janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphIndexTest.java:1137]: Could you please add one more test case to cover the branch mentioned [here](https://github.com/JanusGraph/janusgraph/pull/3203#discussion_r985127606)? I know the branch has been removed, but right now that default case is still not tested. An example would be:
```java
assertEquals(25, tx.traversal().V().or(__.has("name", "bob"), __.has("prop", "val")).values("age").max().next());
```
i.e. A mixed index is matched, but the index itself does not cover the property of interest. In other words, the mixed index is not a covering index.

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/strategy/AbstractJanusGraphMixedIndexAggStrategy.java
Comment (li-boxuan) [inline janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/strategy/AbstractJanusGraphMixedIndexAggStrategy.java:84]: missing param `hasField` in javadoc

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/strategy/JanusGraphMixedIndexAggStrategy.java
Comment (li-boxuan) [inline janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/strategy/JanusGraphMixedIndexAggStrategy.java:61]: Not necessary to be included in this PR, but I guess a good next step is to also add support for `values(key).count()`.

**Review by li-boxuan — COMMENTED** [review]
Thanks for the great contribution! I just have some nitpicks, otherwise LGTM.

File: docs/changelog.md
Comment (To-om) [reply docs/changelog.md:164]: Of course. I will remove it.

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/strategy/JanusGraphMixedIndexAggStrategy.java
Comment (To-om) [reply janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/strategy/JanusGraphMixedIndexAggStrategy.java:61]: Yes I add support of this query. It is equivalent to `has(key).count()` which is optimized.

File: janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphIndexTest.java
Comment (To-om) [reply janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphIndexTest.java:1137]: I've added a [test](https://github.com/StrangeBeeCorp/janusgraph/blob/aggregations/janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/IndexMetricTest.java#L71) that cover the default case, but only for metrics (the result is not checked).
I'll add this test.

**Review by li-boxuan — APPROVED** [review]
LGTM 👍

**Comment by analytically** [comment]
@To-om 

janusgraph-solr/src/main/java/org/janusgraph/diskstorage/solr/SolrIndex.java#L781
Added line #L781 was not covered by tests

**Review by porunov — APPROVED** [review]
LGTM. Thank you for this feature @To-om !

**Comment by porunov** [comment]
@To-om could you please squash your commits into a single commit?
If it's complicated then you can just hard reset from this branch: https://github.com/porunov/janusgraph/tree/aggregations
Alternatively you can click `Allow edits and access to secrets by maintainers` on this PR and I will force push into your branch.

**Pull request merged**
